server/src/main/java/io/crate/statistics/ReservoirSampler.java

```diff
@@ -183,159 +159,211 @@ Samples getSamples(RelationName relationName, List<Reference> columns, int maxSa
+        return new ExpressionsAndInputs(expressions, inputs);
     }
 
-    @VisibleForTesting
-    ArrayList<Row> createRecords(LongArrayList samples,
-                                 List<IntFunction<Object[]>> docIdToRowsFunctionPerReader,
-                                 RamAccounting ramAccounting,
-                                 CellsSizeEstimator sizeEstimator,
-                                 int maxSamples) {
-        ArrayList<Row> records = new ArrayList<>();
-        long bytesSinceLastPause = 0;
-
-        for (LongCursor cursor : samples) {
-            long fetchId = cursor.value;
-            int readerId = FetchId.decodeReaderId(fetchId);
-            IntFunction<Object[]> docIdToRow = docIdToRowsFunctionPerReader.get(readerId);
-            Object[] row = docIdToRow.apply(FetchId.decodeDocId(fetchId));
+    private static class ColumnCollector<T> {
+        LuceneCollectorExpression<?> collector;
+        Input<?> input;
+        final ColumnStatsBuilder<T> statsBuilder;
 
-            try {
-                long bytesRead = sizeEstimator.estimateSize(row);
-                ramAccounting.addBytes(bytesRead);
-                bytesSinceLastPause = maybePause(bytesRead, bytesSinceLastPause);
-            } catch (CircuitBreakingException e) {
-                LOGGER.info(
-                    "Stopped gathering samples for `ANALYZE` operation because circuit breaker triggered. "
-                        + "Generating statistics with {} instead of {} records",
-                    records.size(),
-                    maxSamples
-                );
-                break;
+        private ColumnCollector(ColumnStatsBuilder<T> statsBuilder) {
+            this.statsBuilder = statsBuilder;
+        }
+
+        void setShard(LuceneCollectorExpression<?> collector, Input<?> input) {
+            this.collector = collector;
+            this.input = input;
+        }
+
+        @SuppressWarnings("unchecked")
+        void collect(int docId, RamAccounting ramAccounting) {
+            collector.setNextDocId(docId);
+            statsBuilder.add((T) input.value(), ramAccounting);
+        }
+    }
+
+    private record ExpressionsAndInputs(
+        List<? extends LuceneCollectorExpression<?>> expressions,
+        List<Input<?>> inputs
+    ) {}
+
+    private record ShardExpressions(Engine.Searcher searcher,
+                                    ExpressionsAndInputs expressionsAndInputs) {
+
+        void updateColumnCollectors(List<ColumnCollector<?>> collectors) {
+            assert collectors.size() == expressionsAndInputs.expressions.size();
+            assert collectors.size() == expressionsAndInputs.inputs.size();
+            for (int i = 0; i < collectors.size(); i++) {
+                collectors.get(i).setShard(expressionsAndInputs.expressions.get(i), expressionsAndInputs.inputs.get(i));
             }
-            records.add(new RowN(row));
         }
 
-        return records;
     }
```

Comment:
Would it make sense to re-use `ExpressionsInput` here? Part of this looks similar.

Reply:
This is a great idea, I'll see if I can simplify things out by using the lucene expressions directly.

Reply:
This has simplified things a lot.  I think we will be able to simplify further when we split the ColumnStatsBuilder into a local ColumnSketchBuilder and a serializable ColumnSketch, as we can then make the builder directly extend LuceneCollectorExpression and remove the ColumnSampler class.